Revert "net: tcp: Fix possible deadlock in tcp_conn_unref()"

neja20 opened the pull request:
Reverts zephyrproject-rtos/zephyr#42526

As I am testing this change, it seems that the net_ctx is getting leaked. So it seems it may not be the rightway to unlock the tcp_lock.

I would like to consider the option of manipulating `ctx->cond.lock` lock. For example get rid of `(void)k_mutex_lock(ctx->cond.lock, K_FOREVER);` in `static void zsock_received_cb()`. How do u feel about that? Because it seems to work better than the other way

nashif commented:
you need a signoff in the commit message.

rlubos commented:
> I would like to consider the option of manipulating ctx->cond.lock lock. For example get rid of (void)k_mutex_lock(ctx->cond.lock, K_FOREVER); in static void zsock_received_cb().

I don't think we should remove that mutex protection - it was added for a reason, removing it could lead to other issues with recv queue handling in the socket/net_context.

I'm thinking about the `tcp_conn_unref()` function - it seems to lock the mutex first, and only after that it decrements the ref count with  `atomic_dec()` (which should be safe w/o a mutex protection) and check if it actually needs to release the tcp context (and this part should be protected by the mutex in my understanding).

Maybe a good solution would be to move this check before the mutex lock - only one thread should attempt to lock the tcp mutex then, therefore preventing the deadlock. It's implemented in a similar way in net_context (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/net_context.c#L361).

rlubos commented:
CC @jukkar

neja20 commented:
Thanks @rlubos , that sounds reasonable. I will take a look at that

neja20 commented:
hi,

@rlubos's suggestions led me to implement the `tcp_conn_unref()` similary to https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/net_context.c#L361, which definitely helped and revealed the hanging lock in the sockets.

There is a condition variable `ctx->cond.recv` which I assume is being used as a replacement to Events, but the k_condvar_wait() reaquires the lock after the wait is finished and the lock was never released which was the other reason why could tcp stack's `rx_q` thread be locked infinitely, hence the `added k_mutex_unlock()` call.

rlubos commented on subsys/net/ip/tcp.c:
```diff
@@ -361,8 +361,12 @@ static int tcp_conn_unref_debug(struct tcp *conn, const char *caller, int line)
 {
-	int ref_count = atomic_get(&conn->ref_count);
 	struct net_pkt *pkt;
+	int ref_count = atomic_dec(&conn->ref_count) - 1;
```
And move the decrement and check just before the mutex lock, so that we don't check other logic.

rlubos commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -980,8 +980,13 @@ int zsock_wait_data(struct net_context *ctx, k_timeout_t *timeout)
+		/* Release the lock because the condition variable is
+		 * just used for signaling purpose.
+		 */
+		k_mutex_unlock(ctx->cond.lock);
```
I don't think it's right - one of the purpose of `k_condvar_wait()` is to restore the lock once it quits - the socket lock will be released in the corresponding `recv()`/`recvfrom()` call.

neja20 replied:
You are right that the lock is aquiered when calling `recv()`/`recvfrom()` but when calling `read()` it enters the `k_condvar_wait()` without previously owning the lock so that's where the uneven number of lock/unlock is happening.

Is this behavior corret? Should I only call `recv()` and not `read()` from my app?

rlubos replied:
Ok, that's not right then, `read()` should also acquire the mutex. It seems that this should be addressed in the `fdtable.c` which provides generic `read()` implementation (I think the mutex should be acquired before vtable call, just like in case other calls).

neja20 replied:
These changes in PR works well for me. I dont't know if I should add the locking to other functions working with vtable like `write()`, `close()`?

rlubos replied:
I think it'd make sense, both for consistency and additional protection.

carlescufi commented:
@neja20 you don't need to close the PR, just force-push to your branch instead.

neja20 commented:
@carlescufi I am sorry, I might be using git the wrong way. I will find a way to do this properly next time.

carlescufi commented:
> @carlescufi I am sorry, I might be using git the wrong way. I will find a way to do this properly next time.

The usual workflow is: 

```
<make changes>
$ git add
$ git commit --amend
$ git push --force
```

rlubos commented on subsys/net/ip/tcp.c:
```diff
@@ -377,15 +377,13 @@ static int tcp_conn_unref(struct tcp *conn)
-
 	ref_count = atomic_dec(&conn->ref_count) - 1;
-	if (ref_count) {
-		tp_out(net_context_get_family(conn->context), conn->iface,
-		       "TP_TRACE", "event", "CONN_DELETE");
```
You shouldn't delete these lines.

rlubos reviewed (commented):
One more comment.

rlubos reviewed (approved):
Thanks, looking good.

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -377,15 +377,15 @@ static int tcp_conn_unref(struct tcp *conn)
 		tp_out(net_context_get_family(conn->context), conn->iface,
-		       "TP_TRACE", "event", "CONN_DELETE");
-		goto unlock;
+			"TP_TRACE", "event", "CONN_DELETE");
```
Unnecessary change in this line

neja20 replied:
Fixed

jukkar commented on lib/os/fdtable.c:
```diff
@@ -224,21 +224,37 @@ int z_alloc_fd(void *obj, const struct fd_op_vtable *vtable)
 
```
Typo in commit message of `lib/os: fdtable: add locking to posix api`
s/sokcets/sockets/

neja20 replied:
Fixed

jukkar reviewed (commented):
Looks reasonable, couple of minor nits that could be fixed.

dnejezchleb committed:
net: tcp: Fix possible deadlock in tcp_conn_unref()

This reverts commit e7489d8de71b4b4d888f126e9982722927849086.

And fixes the deadlock by allowing only 1 thread to actualy clean up
the connection when the ref_count is 0.

Signed-off-by: Daniel Nejezchleb <dnejezchleb@hwg.cz>

dnejezchleb committed:
lib/os: fdtable: add locking to posix api

Added locking to posix read(), write(), close()
for additional protection.

In read() missing lock would create uneven calls to locking
mechanism in sockets.c after k_condvar_wait().
That results in socket lock not ever being unlocked

Signed-off-by: Daniel Nejezchleb <dnejezchleb@hwg.cz>
